vendor/parse-wyc/lib/migrate_ast.ml

```diff
@@ -1,47 +1,36 @@
 module Selected_version = Migrate_parsetree.Ast_408
-module Ast_mapper = Selected_version.Ast_mapper
 module Parsetree = Selected_version.Parsetree
 module Asttypes = Selected_version.Asttypes
 
 module Mapper = struct
-  let structure = Selected_version.map_structure
-
-  let signature = Selected_version.map_signature
-
-  (* Missing from ocaml_migrate_parsetree *)
-  let use_file (mapper : Ast_mapper.mapper) use_file =
-    let open Parsetree in
-    List.map
-      (fun toplevel_phrase ->
-        match (toplevel_phrase : toplevel_phrase) with
-        | Ptop_def structure ->
-            Ptop_def (mapper.Ast_mapper.structure mapper structure)
-        | Ptop_dir { pdir_name; pdir_arg; pdir_loc } ->
-            let pdir_arg =
-              match pdir_arg with
-              | None -> None
-              | Some a ->
-                  Some { a with pdira_loc = mapper.location mapper a.pdira_loc }
-            in
-            Ptop_dir
-              {
-                pdir_name =
-                  { pdir_name with loc = mapper.location mapper pdir_name.loc };
-                pdir_arg;
-                pdir_loc = mapper.location mapper pdir_loc;
-              })
-      use_file
-
-  type 'a fragment =
-    | Structure : Parsetree.structure fragment
-    | Signature : Parsetree.signature fragment
-    | Use_file : Parsetree.toplevel_phrase list fragment
-
-  let map_ast (type a) (x : a fragment) : Ast_mapper.mapper -> a -> a =
-    match x with
-    | Structure -> structure
-    | Signature -> signature
-    | Use_file -> use_file
+  type ('omp, 'ppxlib) fragment =
```

Comment:
Why does fragment also have to depend on `'omp` and can't only depend on `'ppxlib` as in lib/Migrate_ast? Is that because there's no module `Parser` in ppxlib?

Reply:
The one in `lib` is used only to apply a mapper to a different part of an AST. This one is also used to convert from the omp type to the ppxlib type (it encodes the relation between `OMP.structure` and `Ppxlib.structure`). There's actually [a very similar one in ppxlib](https://github.com/ocaml-ppx/ppxlib/blob/70ccb695fb0e3833bdddfbf6401900cd32b5d58d/ast/import.ml#L13). It might have been possible to use it instead of redefining one, but that means that `iter_ast` and `entrypoint` (in `parse_wyc.ml`) would have to handle all the cases. I don't know if it's possible to restrict the cases without modifying the type.